cppcheck: Possible nullptr-dereference

I need to fix cppcheck errors. I finally got it working again:

<details>
<summary>Click to see</summary>

```
src/file_search/TvShowFileSearcher.cpp:306:63: warning: Either the condition 'filenameParts.count()>1' is redundant or filenameParts.at() argument nr 1 can have invalid value. The value is -1 but the valid values are '0:'. [invalidFunctionArg]
            filename = filenameParts.at(filenameParts.count() - 3);
                                                              ^
src/file_search/TvShowFileSearcher.cpp:305:35: note: Assuming that condition 'filenameParts.count()>1' is not redundant
        if (filenameParts.count() > 1 && helper::isDvd(files.at(0))) {
                                  ^
src/file_search/TvShowFileSearcher.cpp:306:63: note: Invalid argument
            filename = filenameParts.at(filenameParts.count() - 3);
                                                              ^
src/file_search/TvShowFileSearcher.cpp:356:63: warning: Either the condition 'filenameParts.count()>1' is redundant or filenameParts.at() argument nr 1 can have invalid value. The value is -1 but the valid values are '0:'. [invalidFunctionArg]
            filename = filenameParts.at(filenameParts.count() - 3);
                                                              ^
src/file_search/TvShowFileSearcher.cpp:355:35: note: Assuming that condition 'filenameParts.count()>1' is not redundant
        if (filenameParts.count() > 1 && helper::isDvd(files.at(0))) {
                                  ^
src/file_search/TvShowFileSearcher.cpp:356:63: note: Invalid argument
            filename = filenameParts.at(filenameParts.count() - 3);
                                                              ^
src/media_center/KodiXml.cpp:1404:12: error: syntax error [syntaxError]
    return QStringLiteral();
           ^
src/media/MediaInfoFile.h:19:29: error: Returning object that will be invalid when returning. [returnDanglingLifetime]
    return str.toUtf8().data(); // TODO: Can we use toStdString()?
                            ^
src/media/MediaInfoFile.h:19:29: note: Pointer to container is created here.
    return str.toUtf8().data(); // TODO: Can we use toStdString()?
                            ^
src/media/MediaInfoFile.h:19:29: note: Returning object that will be invalid when returning.
    return str.toUtf8().data(); // TODO: Can we use toStdString()?
                            ^
src/model/tv_show/TvShowUtils.cpp:13:63: warning: Either the condition 'filenameParts.count()>1' is redundant or filenameParts.at() argument nr 1 can have invalid value. The value is -1 but the valid values are '0:'. [invalidFunctionArg]
            filename = filenameParts.at(filenameParts.count() - 3);
                                                              ^
src/model/tv_show/TvShowUtils.cpp:12:35: note: Assuming that condition 'filenameParts.count()>1' is not redundant
        if (filenameParts.count() > 1 && helper::isDvd(files.first())) {
                                  ^
src/model/tv_show/TvShowUtils.cpp:13:63: note: Invalid argument
            filename = filenameParts.at(filenameParts.count() - 3);
                                                              ^
src/scrapers/movie/tmdb/TmdbMovie.h:29:10: style: Virtual function 'initialize' is called from constructor 'TmdbMovie(QObject*parent=nullptr)' at line 165. Dynamic binding is not used. [virtualCallInConstructor]
    void initialize() override;
         ^
src/scrapers/movie/tmdb/TmdbMovie.cpp:165:5: note: Calling initialize
    initialize();
    ^
src/scrapers/movie/tmdb/TmdbMovie.h:29:10: note: initialize is a virtual function
    void initialize() override;
         ^
src/ui/media_center/KodiSync.cpp:550:5: warning: Either the condition 'reply==nullptr' is redundant or there is possible null pointer dereference: reply. [nullPointerRedundantCheck]
    reply->deleteLater();
    ^
src/ui/media_center/KodiSync.cpp:551:15: note: Assuming that condition 'reply==nullptr' is not redundant
    if (reply == nullptr) {
              ^
src/ui/media_center/KodiSync.cpp:550:5: note: Null pointer dereference
    reply->deleteLater();
    ^
src/ui/movies/MovieWidget.cpp:883:34: warning: Either the condition 'm_movie==nullptr' is redundant or there is possible null pointer dereference: m_movie. [nullPointerRedundantCheck]
    if ((m_movie == nullptr) && !m_movie->hasLocalTrailer()) {
                                 ^
src/ui/movies/MovieWidget.cpp:883:18: note: Assuming that condition 'm_movie==nullptr' is not redundant
    if ((m_movie == nullptr) && !m_movie->hasLocalTrailer()) {
                 ^
src/ui/movies/MovieWidget.cpp:883:34: note: Null pointer dereference
    if ((m_movie == nullptr) && !m_movie->hasLocalTrailer()) {
                                 ^
src/ui/concerts/ConcertWidget.cpp:72:100: warning: Redundant assignment of 'ui' to itself. [selfAssignment]
    connect(ui->concertStreamdetails, &ConcertStreamDetailsWidget::runtimeChanged,       this, [ui = ui](std::chrono::minutes runtime) {
                                                                                                   ^
src/ui/tv_show/TvShowWidgetEpisode.cpp:496:36: warning: Either the condition 'm_episode!=nullptr' is redundant or there is possible null pointer dereference: m_episode. [nullPointerRedundantCheck]
    StreamDetails* streamDetails = m_episode->streamDetails();
                                   ^
src/ui/tv_show/TvShowWidgetEpisode.cpp:486:20: note: Assuming that condition 'm_episode!=nullptr' is not redundant
    if ((m_episode != nullptr) && m_episode->isDummy()) {
                   ^
src/ui/tv_show/TvShowWidgetEpisode.cpp:496:36: note: Null pointer dereference
    StreamDetails* streamDetails = m_episode->streamDetails();
                                   ^
```

</details>

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -41,12 +41,6 @@ pipeline {
               '''
           }
         }
-        stage('cppcheck') {
-          steps {
-            // Currently allow cppcheck to fail due to false positives
-            sh './scripts/run_cppcheck.sh || echo "Cppcheck Failed"'
-          }
-        }
       }
     }
 
@@ -58,6 +52,14 @@ pipeline {
           '''
       }
     }
+
+    // Requires MOC files created during build
+    stage('cppcheck') {
+      steps {
+        sh './scripts/run_cppcheck.sh'
+      }
+    }
+
     stage('Test') {
       steps {
         // Because our tests require a GUI (even unit tests at the moment) we need a

--- a/scripts/run_cppcheck.sh
+++ b/scripts/run_cppcheck.sh
@@ -16,8 +16,9 @@ print_important "Run cppcheck on all source files"
 
 if [[ ! -f ./build/ci/compile_commands.json ]]; then
 	echo ""
-	print_important "compile_command.json not found, running 'cmake --preset ci'"
+	print_important "compile_command.json not found, running 'cmake --preset ci' and building it for Qt moc-files"
 	cmake --preset ci
+	cmake --build --preset ci
 fi
 
 print_info "Will only print cppcheck warnings"

--- a/scripts/run_cppcheck.sh
+++ b/scripts/run_cppcheck.sh
@@ -23,9 +23,9 @@ fi
 
 print_info "Will only print cppcheck warnings"
 
-# TODO: performance
+# TODO: performance + enable some suppressed checks
 # Notes: ignore quazip, use qt.cfg
-cppcheck --enable=warning,portability,information,missingInclude \
+cppcheck --enable=style,warning,portability,information,missingInclude \
 	--error-exitcode=1 \
 	--inline-suppr \
 	-q \
@@ -33,6 +33,26 @@ cppcheck --enable=warning,portability,information,missingInclude \
 	-j4 \
 	--suppress=preprocessorErrorDirective \
 	--suppress=ignoredReturnValue \
+    --suppress=ignoredReturnErrorCode \
+    --suppress=useStlAlgorithm \
+    --suppress=useInitializationList \
+    --suppress=qSortCalled \
+    --suppress=constParameter \
+    --suppress=constVariable \
+    --suppress=passedByValue \
+    --suppress=noExplicitConstructor \
+    --suppress=shadowFunction \
+    --suppress=unreadVariable \
+    --suppress=knownConditionTrueFalse \
+    --suppress=duplicateBreak \
+    --suppress=unsafeClassCanLeak \
+    --suppress=duplicateCondition \
+    --suppress=redundantInitialization \
+    --suppress=cstyleCast \
+    --suppress=qSwapCalled \
+    --suppress=variableScope \
+    --suppress=unusedVariable \
+    --suppress=constParameterCallback \
 	--inline-suppr \
 	--std=c++14 \
 	--library=qt \

--- a/scripts/run_cppcheck.sh
+++ b/scripts/run_cppcheck.sh
@@ -44,7 +44,6 @@ cppcheck --enable=style,warning,portability,information,missingInclude \
     --suppress=shadowFunction \
     --suppress=unreadVariable \
     --suppress=knownConditionTrueFalse \
-    --suppress=duplicateBreak \
     --suppress=unsafeClassCanLeak \
     --suppress=duplicateCondition \
     --suppress=redundantInitialization \

--- a/src/file_search/TvShowFileSearcher.cpp
+++ b/src/file_search/TvShowFileSearcher.cpp
@@ -302,7 +302,8 @@ SeasonNumber TvShowFileSearcher::getSeasonNumber(QStringList files)
     QStringList filenameParts = files.at(0).split('/');
     QString filename = filenameParts.last();
     if (filename.endsWith("VIDEO_TS.IFO", Qt::CaseInsensitive)) {
-        if (filenameParts.count() > 1 && helper::isDvd(files.at(0))) {
+        // TODO: Re-check: count() > 2? What does the filepath look like?
+        if (filenameParts.count() > 2 && helper::isDvd(files.at(0))) {
             filename = filenameParts.at(filenameParts.count() - 3);
         } else if (filenameParts.count() > 2 && helper::isDvd(files.at(0), true)) {
             filename = filenameParts.at(filenameParts.count() - 2);
@@ -352,7 +353,8 @@ QVector<EpisodeNumber> TvShowFileSearcher::getEpisodeNumbers(QStringList files)
     QStringList filenameParts = files.at(0).split('/');
     QString filename = filenameParts.last();
     if (filename.endsWith("VIDEO_TS.IFO", Qt::CaseInsensitive)) {
-        if (filenameParts.count() > 1 && helper::isDvd(files.at(0))) {
+        // TODO: Re-check: count() > 2? What does the filepath look like?
+        if (filenameParts.count() > 2 && helper::isDvd(files.at(0))) {
             filename = filenameParts.at(filenameParts.count() - 3);
         } else if (filenameParts.count() > 2 && helper::isDvd(files.at(0), true)) {
             filename = filenameParts.at(filenameParts.count() - 2);

--- a/src/media/MediaInfoFile.h
+++ b/src/media/MediaInfoFile.h
@@ -16,13 +16,13 @@ class MediaInfo;
 template<typename T>
 static typename std::enable_if_t<std::is_same<char, T>::value, std::string> toMediaInfoString(const QString& str)
 {
-    return str.toUtf8().data(); // TODO: Can we use toStdString()?
+    return std::string(str.toUtf8().data()); // TODO: Can we use toStdString()?
 }
 
 template<typename T>
 static typename std::enable_if_t<std::is_same<wchar_t, T>::value, std::wstring> toMediaInfoString(const QString& str)
 {
-    return str.toStdWString();
+    return std::wstring(str.toStdWString());
 }
 
 template<typename T>

--- a/src/media_center/KodiXml.cpp
+++ b/src/media_center/KodiXml.cpp
@@ -1401,7 +1401,7 @@ QString saveDataFiles(mediaelch::DirectoryPath basePath,
             return basePath.filePath(file);
         }
     }
-    return QStringLiteral();
+    return {};
 }
 
 QString

--- a/src/model/ActorModel.cpp
+++ b/src/model/ActorModel.cpp
@@ -54,10 +54,7 @@ QVariant ActorModel::data(const QModelIndex& index, int role) const
     switch (role) {
     case ActorRoles::ActorRole: return QVariant::fromValue(actor);
 
-    case ActorRoles::ImageRole: {
-        return actor->image;
-        break;
-    }
+    case ActorRoles::ImageRole: return actor->image;
 
     case Qt::DisplayRole:
     case Qt::EditRole: {

--- a/src/model/tv_show/TvShowUtils.cpp
+++ b/src/model/tv_show/TvShowUtils.cpp
@@ -9,7 +9,8 @@ QString guessTvShowTitleFromFiles(const FileList& files)
     const QStringList filenameParts = files.first().toString().split('/');
     QString filename = filenameParts.last();
     if (filename.endsWith("VIDEO_TS.IFO", Qt::CaseInsensitive)) {
-        if (filenameParts.count() > 1 && helper::isDvd(files.first())) {
+        // TODO: Re-check: count() > 2? What does the filepath look like?
+        if (filenameParts.count() > 2 && helper::isDvd(files.first())) {
             filename = filenameParts.at(filenameParts.count() - 3);
         } else if (filenameParts.count() > 2 && helper::isDvd(files.first(), true)) {
             filename = filenameParts.at(filenameParts.count() - 2);

--- a/src/scrapers/movie/tmdb/TmdbMovie.h
+++ b/src/scrapers/movie/tmdb/TmdbMovie.h
@@ -16,7 +16,7 @@
 namespace mediaelch {
 namespace scraper {
 
-class TmdbMovie : public MovieScraper
+class TmdbMovie final : public MovieScraper
 {
     Q_OBJECT
 public:

--- a/src/ui/concerts/ConcertWidget.cpp
+++ b/src/ui/concerts/ConcertWidget.cpp
@@ -69,8 +69,8 @@ ConcertWidget::ConcertWidget(QWidget* parent) : QWidget(parent), ui(new Ui::Conc
         onInfoChanged();
     });
     connect(ui->buttonRevert,         &QAbstractButton::clicked,                         this, &ConcertWidget::onRevertChanges);
-    connect(ui->concertStreamdetails, &ConcertStreamDetailsWidget::runtimeChanged,       this, [ui = ui](std::chrono::minutes runtime) {
-        ui->concertInfo->setRuntime(runtime);
+    connect(ui->concertStreamdetails, &ConcertStreamDetailsWidget::runtimeChanged,       this, [ui_ = ui](std::chrono::minutes runtime) {
+        ui_->concertInfo->setRuntime(runtime);
     });
     // clang-format on
 

--- a/src/ui/main/Update.cpp
+++ b/src/ui/main/Update.cpp
@@ -42,6 +42,7 @@ void Update::onCheckFinished()
     auto* reply = dynamic_cast<QNetworkReply*>(QObject::sender());
     if (reply == nullptr) {
         qCCritical(generic) << "[Updater] Dynamic Cast Failed";
+        return;
     }
     reply->deleteLater();
     if (reply->error() != QNetworkReply::NoError) {

--- a/src/ui/media_center/KodiSync.cpp
+++ b/src/ui/media_center/KodiSync.cpp
@@ -547,15 +547,16 @@ void KodiSync::triggerReload()
 void KodiSync::onScanFinished()
 {
     auto* reply = dynamic_cast<QNetworkReply*>(QObject::sender());
-    reply->deleteLater();
     if (reply == nullptr) {
         qCCritical(generic) << "[KodiSync] dynamic_cast<QNetworkReply*> failed for onScanFinished!";
 
     } else if (reply->error() != QNetworkReply::NoError) {
+        reply->deleteLater();
         ui->status->setText(tr("Error: %1").arg(reply->errorString()));
         qCWarning(generic) << "[KodiSync] Network Error:" << reply->errorString() << "|" << reply->url();
 
     } else {
+        reply->deleteLater();
         ui->status->setText(tr("Finished. Kodi is now loading your updated items."));
     }
 

--- a/src/ui/movies/MovieWidget.cpp
+++ b/src/ui/movies/MovieWidget.cpp
@@ -880,7 +880,7 @@ void MovieWidget::onDownloadTrailer()
 
 void MovieWidget::onPlayLocalTrailer()
 {
-    if ((m_movie == nullptr) && !m_movie->hasLocalTrailer()) {
+    if (m_movie == nullptr || !m_movie->hasLocalTrailer()) {
         return;
     }
 

--- a/src/ui/tv_show/TvShowWidgetEpisode.cpp
+++ b/src/ui/tv_show/TvShowWidgetEpisode.cpp
@@ -317,7 +317,7 @@ void TvShowWidgetEpisode::onClear()
  */
 void TvShowWidgetEpisode::onSetEnabled(bool enabled)
 {
-    if ((m_episode != nullptr) && m_episode->isDummy()) {
+    if (m_episode != nullptr && m_episode->isDummy()) {
         ui->groupBox_3->setEnabled(false);
         return;
     }
@@ -483,7 +483,7 @@ void TvShowWidgetEpisode::updateStreamDetails(bool reloadedFromFile)
 {
     Q_UNUSED(reloadedFromFile);
 
-    if ((m_episode != nullptr) && m_episode->isDummy()) {
+    if (m_episode == nullptr || m_episode->isDummy()) {
         return;
     }
 

